feat: location import (spreadsheet validation)

shafy opened the pull request:
This is the spreadsheet validation part of the location import feature. It adds a spreadsheet (we're using [jexcel](https://github.com/jspreadsheet/jexcel)) and validates the data when the user clicks "Start import". Notably, it highlights the rows and cells that are erroneous and displays detailed error messages below the spreadsheet.

This doesn't create any locations in the database. Once we have the background job that actually creates the locations, passing validations will kick off the job.

One caveat is that for opening time errors, we can't highlight specific cells because Rails validation doesn't provide the exact info. Therefore, we just highlight all opening times cells for now. We could later add some logic that checks the validity of each `opening_time` one by one, but I decided to skip this for now.

Here's a short demo:
https://user-images.githubusercontent.com/1881676/164773449-7e25660a-0dda-4948-90c0-35857d734736.mov

Here's how the detailed error message look:
<img width="1360" alt="Screenshot 2022-04-22 at 20 05 57" src="https://user-images.githubusercontent.com/1881676/164773554-7e835093-32b5-4f83-a99e-0d3461af9119.png">

To-Do:
- [x] Finish removing CanCanCanCanCan
- [x] Add tests
- [x] Fix weird `before_action` bug in `/dashboard/locations_controller.rb`

sagzy reviewed (commented):
Code is super clean. Nothing to add. Fixed the before_action issue in [9c91f84](https://github.com/mapzy/mapzy/pull/68/commits/9c91f843cb14c1c47137bd4d679d72ecd7d2c2bf).

Feedback on the product side:
- The spreadsheet is very smooth to use. The instructions are clear. Just feels great to play with it! 💯
- The error highlighting is a killer feature. Makes it much nicer to handle data issues

- Colors: the red is both used for the CTA & to highlight errors. Whereas the flash message for errors is purple. It's a bit weird. It would make more sense to have a single color for the errors (the red) and a different color for the CTA. 

- Instructions: pretty clear and very readable. There is a mix of italic and "" in the copyright that is not consistent.

- Instructions UI: the instruction box quite some space on the screen, imo. It's something the user needs to read once or twice and having more space for the spreadsheet is more important. I would reduce at least reduce the font size and maybe put them into a sort of "Notion callout"
<img width="743" alt="Screenshot 2022-04-23 at 10 27 05" src="https://user-images.githubusercontent.com/6225080/164886820-4e0d081c-26ba-4366-b84b-e6142b85daad.png">

- Error messages: there are now two flash messages when there is an error. One of the very top of the page that is generic and one at the bottom that is more specific. Would keep only one. Best would be to have the error message as close as possible to the excel container

- General structure of the layout: the current structure from top to bottom is: generic error message, title, instructions, CTA, excel, specific error message. It looks a bit unnatural to me, what about following a more conventional structure, e.g. title, instructions, excel with eventual errors, CTA? 

Bugs:
- If I leave name or address blank, I get the expected error and highlight of the error. But even if I fix it, it doesn't disappear. I noticed that the url is `/location_imports` instead of `/location_imports/new` after an error. Maybe linked to this?

https://user-images.githubusercontent.com/6225080/164887167-8d3e156f-ca23-43d7-ae9f-c3c623eade9d.mov

shafy commented on app/controllers/dashboard/location_imports_controller.rb:
```diff
@@ -0,0 +1,56 @@
+          "", ""                # sunday
+        ], [], [], [], [], [], [], [], [], [] # empty rows
+      ]
+    end
```
No, we don't ignore it. The user is supposed to paste over it with their own data. Do you think this is not clear?

sagzy replied:
Yeah it wasn't super clear to me. But maybe just me

shafy commented:
I've updated the design a bit, incorporating your comments:
<img width="1440" alt="Screenshot 2022-04-25 at 19 14 53" src="https://user-images.githubusercontent.com/1881676/165159016-afbfc12a-3ad2-4572-8074-b13dc9332113.png">


- I want to keep the CTA button red (that's how it also looks across Mapzy), so I changed the errors to the blueish color.
- Improved the UI a bit with the Notion stlye box, import button at the end, only one error box etc.
- I've also added the mass import. Since we skip validations (performance, and it's already validated at that point), it automatically inserts `nil` for long and lat.
- For the mass import, I also changed the logic of the `LocationImport` model a bit (before, it was like one `LocationImport` model per `Location`, now it's one `LocationImport` model that holds all `Locations` - which makes more sense and also is better code.

sagzy commented:
> I've also added the mass import. Since we skip validations (performance, and it's already validated at that point), it automatically inserts nil for long and lat.

It should be `0` for long and lat, so that it's considered `pending`. Is that what you meant?

shafy commented:
> > I've also added the mass import. Since we skip validations (performance, and it's already validated at that point), it automatically inserts nil for long and lat.
> 
> It should be `0` for long and lat, so that it's considered `pending`. Is that what you meant?

Can it also be `nil`? That's how it is currently.

sagzy commented:
UI: looks much better, thanks! 

Instructions:
- we can remove "(up to 500 per import)" in the instructions. There is no limit in the geocoding task at least.
- _Description_ is optional => "Description" is optional and 18:00 for 6 pm => _18:00_ for 6 pm (to be consistent in the italic / "" styles)

sagzy commented:
> > > I've also added the mass import. Since we skip validations (performance, and it's already validated at that point), it automatically inserts nil for long and lat.
> > 
> > 
> > It should be `0` for long and lat, so that it's considered `pending`. Is that what you meant?
> 
> Can it also be `nil`? That's how it is currently.

No, then it would be considered as `geocoding_error` instead of `geocoding_pending`

shafy commented:
> > > > I've also added the mass import. Since we skip validations (performance, and it's already validated at that point), it automatically inserts nil for long and lat.
> > > 
> > > 
> > > It should be `0` for long and lat, so that it's considered `pending`. Is that what you meant?
> > 
> > 
> > Can it also be `nil`? That's how it is currently.
> 
> No, then it would be considered as `geocoding_error` instead of `geocoding_pending`

Ok let me change that.
